datastore/store.go

```diff
@@ -51,13 +51,12 @@ func (db *CachedDB) GetAtxHeader(id types.ATXID) (*types.ActivationTxHeader, err
 		return nil, fmt.Errorf("get ATXs from DB: %w", err)
 	}
 
-	db.atxHdrCache.Add(id, &atx.ActivationTxHeader)
-	return &atx.ActivationTxHeader, nil
+	return atx.Header(), nil
```

Comment:
here you are not getting it from the cache anymore. you are creating a new struct?

Reply:
it adds to the cache in `GetFullAtx`, but you are not returning the cached value. you are creating a new struct. what am i missing?

Reply:
also, what is the value of the header cache if one can arbitrarily call atx.Header() now?
should Header() be a private method in the cache instead?

Reply:
I didn't change any behaviour. The code behaves the same as before:

- If you call either `GetFullAtx(id)` or `GetAtxHeader(id)` you will receive the Atx / header for the given ID
- In the case of `GetAtxHeader` if the value for the id is cached the cached value is returned
- If `GetAtxHeader(id)` is called and the value is cached, the cached value is returned
- If `GetFullAtx(id)` or `GetAtxHeader(id)` are called when the header for the given ID is not cached it will be cached

The only thing that changed is: if a header isn't cached for a given ID it is now only added to the cache once instead of twice.

Reply:
not what my reading.
old code:
```
func (db *CachedDB) GetAtxHeader(id types.ATXID) (*types.ActivationTxHeader, error) {
	if id == *types.EmptyATXID {
		return nil, errors.New("trying to fetch empty atx id")
	}

	if atxHeader, gotIt := db.atxHdrCache.Get(id); gotIt {
		return atxHeader, nil
	}
	atx, err := db.GetFullAtx(id)
	if err != nil {
		return nil, fmt.Errorf("get ATXs from DB: %w", err)
	}

	db.atxHdrCache.Add(id, &atx.ActivationTxHeader)
	return &atx.ActivationTxHeader, nil
}
```
it cached the same ptr twice (once in `GetFullAtx()`, and once in `GetAtxHeader()`)
but it is returning the cached ptr. you are now returning a new struct.

Reply:
> It's the same as before: one could have arbitrarily called atx.ActivationTxHeader instead of fetching the header from the cache and Header() can't be a private method because it is required to be called in packages other than common/types

no. before all the header ptrs are cached after retrieving the full atx from db. here you are creating a new struct every time Header() is called.

Reply:
I get your point, but what is the behaviour we are expecting?

```golang
id := 1234
header1 := db.GetAtxHeader(id)

// make some change
header1.PrevATXID = 5678

header2 := db.GetAtxHeader(id)
equal := reflect.DeepEqual(header1, header2)
```

Is equal expected to be true or false?

Reply:
i don't understand your example. i don't think that's how the codebase use the atx header cache....

the expected behavior is caching the header, not creating more memory than we need to.
now if you want to argue the need of the cache, that's a different matter.

Reply:
I removed the `Header()` method from `VerifiedActivationTx`. Now a header can only be fetched from the store/cache and is cached correctly after fetching it (or the full ATX with the same ID).